compiler/src/dotty/tools/dotc/reporting/WConf.scala

```diff
@@ -5,13 +5,20 @@ package reporting
 import scala.language.unsafeNulls
 
 import dotty.tools.dotc.core.Contexts._
-import dotty.tools.dotc.util.SourcePosition
+import dotty.tools.dotc.util.{NoSourcePosition, SourcePosition}
+import dotty.tools.dotc.interfaces.SourceFile
+import dotty.tools.dotc.reporting.MessageFilter.SourcePattern
 
 import java.util.regex.PatternSyntaxException
 import scala.annotation.internal.sharable
 import scala.util.matching.Regex
+import scala.jdk.OptionConverters._
+import scala.collection.mutable
 
 enum MessageFilter:
+
+  private val sourcePatternCache = mutable.Map.empty[SourceFile, Boolean]
```

Comment:
The cache should probably be cleared on each compiler run (`dotc.Compiler.newRun`). What utilities does dotty have to manage caches?

Reply:
There is the `PerRun` in `Definitions`. 

Or we could make it a Map of (pattern, SourceFile) -> Boolean. I think there is no state to influence the check if the source file matches a pattern.

Reply:
Multiple compiler instances running at the same time might concurrently try to access and update this cache, so if this really needs to be a global map (seems like it should be tied to the Context instead?) it needs to be a ConcurrentHashMap (I'm surprised this wasn't caught by https://github.com/lampepfl/dotty/blob/main/compiler/src/dotty/tools/dotc/transform/CheckReentrant.scala)

Reply:
Not knowing how this works exactly from the concurrency perspective I attempted porting over Scala2 implementation which was a plain mutable map. Given concerns about the concurrency though, would it be acceptable to not do caching in this PR? We could always make an optimization later without holding back the feature

Reply:
TBH, it was me who mentioned the cache. I can withdraw my comment

Reply:
cache removed